IS-07-02: Add initial test suite for integration tests between IS-07, IS-04 and IS-05

cristian-recoseanu opened the pull request:
This creates a new test suite IS-07-02 which will be handling all of the integration tests between IS-07, IS-04 and IS-05. Transport tests will be added here as well.

Currently just 2 tests have been added:

- test 01 - checks the ext parameters exist (ext_is_07_source_id, ext_is_07_rest_api_url)
- test 02 - checks if the sources have a corresponding IS-05 sender and if not return a warning

garethsb-sony commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,94 @@
+        if len(self.senders) > 0:
+            for sender in self.senders:
+                dest = "single/senders/" + sender + "/active/"
+                valid, response = self.is05_utils.checkCleanRequestJSON("GET", dest)
```
Do you need to get _all_ the IS-05 Senders, or only the ones whose transport_type is MQTT or WebSocket?

cristian-recoseanu replied:
I think it makes sense to use the IS05Utils wherever possible and then filter.

garethsb-sony commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,94 @@
+
+        if len(self.senders) > 0:
+            for sender in self.senders:
+                dest = "single/senders/" + sender + "/active/"
```
Probably better to use the preferred form without the trailing slash?

cristian-recoseanu replied:
Yep agreed. Changes incoming to support this.

garethsb-sony commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,94 @@
+                dest = "single/senders/" + sender + "/active/"
+                valid, response = self.is05_utils.checkCleanRequestJSON("GET", dest)
+                if valid:
+                    if len(response) > 0 and isinstance(response["transport_params"][0], dict):
```
I guess we don't want an exception at this point, so all the extra checks are needed... hmmm.

garethsb-sony commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,94 @@
+                        if sorted(params) == sorted(extParams):
+                            valid_params = True
+                    elif self.transport_types[sender] == "urn:x-nmos:transport:mqtt":
+                        if sorted(params) == sorted(extParams):
```
Is `ext_is07_source_id` used with MQTT? I thought not, according to the spec.

cristian-recoseanu replied:
No you are correct. Must have missed that. I will make changes to split them up for each transport.

andrewbonney commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,97 @@
+                            if sorted(params) == sorted(ext_params_mqtt):
+                                valid_params = True
+                        if not valid_params:
+                            return test.FAIL("Missing required ext parameters")
```
Might be useful to indicate the affected Sender here to assist with debug:
```suggestion
                            return test.FAIL("Missing required ext parameters for Sender {}".format(sender))
```

cristian-recoseanu replied:
Yep will make the change.

andrewbonney commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,97 @@
+        else:
+            return test.UNCLEAR("Not tested. No resources found.")
+
+        return test.PASS()
```
Presumably this return statement could be removed?

cristian-recoseanu replied:
I think you are right. Will make the change.

andrewbonney commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,97 @@
+                if not found_source:
+                    return test.FAIL("Source {} has no associated IS-05 sender".format(source))
+
+        return test.PASS()
```
Might be best to return 'UNCLEAR' if there are no Sources found?

cristian-recoseanu replied:
Yes in similar fashion to test_01. I will make the change.

garethsb-sony commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -45,30 +45,33 @@ def set_up_tests(self):
 
     def test_01(self, test):
-        """Each Sender has ext_is_07_source_id and ext_is_07_rest_api_url parameters"""
+        """Each Sender has the required ext parameters"""
```
Hmm, is there a bit of a future problem here? When another spec starts using WebSocket or MQTT, this test will fail because that spec's senders won't have the IS-07 `ext_` params...

cristian-recoseanu replied:
Hopefully now fixed by including IS-04 and only testing on true IS-07 resources

cristian-recoseanu commented on nmostesting/IS07Utils.py:
```diff
@@ -0,0 +1,38 @@
+                for sub_path in ["state", "type"]:
+                    valid_sub, sub = TestHelper.do_request("GET", "{}/{}/{}".format(sources_url, source_id, sub_path))
+                    if not valid_sub or sub.status_code != 200:
+                        pass
```
Yup. Well spotted. I think checking for valid is enough.

garethsb-sony replied:
I think the check of `sub.status_code` is a good idea, it just needs to be `continue` instead of `pass`, to move onto the next `sub_path` (or out of that loop).

garethsb-sony commented on nmostesting/IS07Utils.py:
```diff
@@ -32,7 +32,6 @@ def get_sources_states_and_types(self):
-                    if not valid_sub or sub.status_code != 200:
-                        pass
-                    toReturn[source_id][sub_path] = sub.json()
+                    if valid_sub:
```
You can do it this way, rather than with `continue`, but you should still check `and sub.status == 200`,  I think. In order to avoid putting error JSON into the results? Or we have to handle that everywhere.

cristian-recoseanu replied:
Ok. That makes sense.

garethsb-sony commented on nmostesting/IS07Utils.py:
```diff
@@ -32,6 +32,6 @@ def get_sources_states_and_types(self):
                 for sub_path in ["state", "type"]:
                     valid_sub, sub = TestHelper.do_request("GET", "{}/{}/{}".format(sources_url, source_id, sub_path))
-                    if valid_sub:
+                    if valid_sub and sub.status == 200:
```
Isn't it `status_code`?

cristian-recoseanu replied:
Yes it is. Copy paste error.

garethsb-sony commented on nmostesting/IS04Utils.py:
```diff
@@ -12,10 +12,54 @@
     def __init__(self, url):
         NMOSUtils.__init__(self, url)
+
+    def get_node(self):
```
`IS04Utils` is also used with IS-04 Registration API and Query API.

Therefore, I think these functions need to be carefully named. Maybe the simplest change is just to change `get_node` to `get_self` which makes it clearer it works with the Node API. The other functions would work with the Query API at least.

cristian-recoseanu replied:
Ok understood. That makes sense.

garethsb-sony commented on nmostesting/IS04Utils.py:
```diff
@@ -12,10 +12,54 @@
+        valid_resource, resource = TestHelper.do_request("GET", self.url + "self")
+        if valid_resource and resource.status_code == 200:
+            return resource.json()
+
```
Does flake8 warn if you put an explicit `return None` in (for the error case)?

cristian-recoseanu replied:
Do you mean something like this?

```
        valid_resource, resource = TestHelper.do_request("GET", self.url + "self")
        if valid_resource and resource.status_code == 200:
            return resource.json()
        else:
            return None
```

garethsb-sony replied:
Yes.

cristian-recoseanu replied:
Ok yep. WIll do

garethsb-sony commented on nmostesting/IS04Utils.py:
```diff
@@ -12,10 +12,54 @@
+
+        return self.get_resources(self.url + "receivers")
+
+    def get_resources(self, url):
```
What do you think about defaulting to `self.url` here?

E.g.

```Python
def get_receivers(self, url=None):

    return self.get_resources("receivers", url)

def get_resources(self, resource, url=None):

    if not url:
        url = self.url
    ... ... TestHelper.do_request("GET", url + resource)
```

cristian-recoseanu replied:
Yep. I think it makes sense. Then they can be used with the Query API if need be.

cristian-recoseanu commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,119 @@
+        self.sources_to_test = {}
+
+        for sender in self.is04_senders:
+            for flow in self.is04_flows:
```
Yep. No point in looping through. Just check it exists in the dictionary. I will make the change

cristian-recoseanu commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,119 @@
+        for sender in self.is04_senders:
+            for flow in self.is04_flows:
+                if flow == self.is04_senders[sender]["flow_id"]:
+                    for source in self.is04_sources:
```
Yep. No point in looping through. Just check it exists in the dictionary. I will make the change

cristian-recoseanu commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,119 @@
+        ext_params_mqtt = ['ext_is_07_rest_api_url']
+        if len(self.senders_to_test.keys()) > 0:
+            for sender in self.senders_to_test:
+                if sender in self.sender_active_params:
```
Would the message be no parameters founds for sender?

garethsb-sony replied:
Yes, something like `"Sender {} not found in Connection API".format(sender)` perhaps?

cristian-recoseanu replied:
Yep looks good. I will make the change

cristian-recoseanu commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,119 @@
+                            return test.FAIL("Source {} has no associated IS-05 sender".format(source))
+                return test.PASS()
+            else:
+                return test.UNCLEAR("Not tested. No resources found.")
```
Instead of UNCLEAR?

garethsb-sony replied:
Yes, since we have IS-07 Sources but no IS-04 Sources to test. Perhaps with a message like `"Source {} not found in Node API".format(self.is07_sources[0])`? (Not quite sure of the syntax there.)

cristian-recoseanu replied:
Ok. Understood

andrewbonney commented:
As this has ended up requiring IS-04 too, is it worth renaming it to IS-07-02 and just using the one file for all IS-04/05 testing?

garethsb-sony commented:
> As this has ended up requiring IS-04 too, is it worth renaming it to IS-07-02 and just using the one file for all IS-04/05 testing?

Yep, Cristian and I think so, though it may involve some test case renumbering when we introduce the tests that only use IS-04 since it feels like they ought to be nearer the top.

cristian-recoseanu commented:
> 
> 
> > As this has ended up requiring IS-04 too, is it worth renaming it to IS-07-02 and just using the one file for all IS-04/05 testing?
> 
> Yep, Cristian and I think so, though it may involve some test case renumbering when we introduce the tests that only use IS-04 since it feels like they ought to be nearer the top.

Yep I think there will always be a renumbering exercise once we have all of our tests together. The transport tests should definately be near the end.

cristian-recoseanu commented:
> 
> 
> As this has ended up requiring IS-04 too, is it worth renaming it to IS-07-02 and just using the one file for all IS-04/05 testing?

Yep I think that makes sense and I think we all agree on that.

cristian-recoseanu commented:
Ok. I have renamed the suite to IS0702.
I have also renamed the PR and my first comment for consistency.
Hopefully all of the comments have been addressed. If not let me know.
Thanks!

garethsb-sony commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,119 @@
+                                                     .format(sender, e))
+                        if not found_source:
+                            return test.FAIL("Source {} has no associated IS-05 sender".format(source))
+                return test.PASS()
```
Do you think we should we add the `else` here too?

```Python
                    else:
                        return test.FAIL("Source {} not found in Node API".format(source))
```

garethsb-sony commented on nmostesting/suites/IS0703Test.py:
```diff
@@ -0,0 +1,119 @@
+                            return test.FAIL("Source {} has no associated IS-05 sender".format(source))
+                return test.PASS()
+            else:
+                return test.FAIL("No sources found in IS-04 Node API")
```
Then we could remove this one?

cristian-recoseanu replied:
Made the change. I hope that's what you had in mind.

garethsb-sony replied:
Yep!

cristian-recoseanu commented:
> 
> 
> I think this is suitable to merge; good work, Cristian!
> 
> One thing I notice is that `test_02` passes so long as there is _at least one_ IS-04 sender of the IS-07 source.
> This is great for implementations that only implement _either_ WebSocket _or_ MQTT, but as written, the spec seems to require both?

I think that is reasonable as a basic test for "do I have at least a sender".
The way the IS-07 spec was thought out was that implementers should always have both options available but in practice we thought end users will never want to run both side by side in the same system. I think they will go for one or the other.

Also I think further transport tests for both MQTT and WebSocket will highlight if there are missing senders when configured for the right transport.

garethsb-sony commented:
> Also I think further transport tests for both MQTT and WebSocket will highlight if there are missing senders when configured for the right transport.

As discussed in Slack, I agree, but I wonder if we ought to have an explicit `test.MANUAL` test case, with a message like `"Implementation is capable of offering both WebSocket and MQTT transports"` and a link to https://github.com/AMWA-TV/nmos-event-tally/blob/v1.0.x/docs/5.0.%20Transports.md#transports.

We can add that after this PR is approved and merged.

cristian-recoseanu commented:
> 
> 
> > Also I think further transport tests for both MQTT and WebSocket will highlight if there are missing senders when configured for the right transport.
> 
> As discussed in Slack, I agree, but I wonder if we ought to have an explicit `test.MANUAL` test case, with a message like `"Implementation is capable of offering both WebSocket and MQTT transports"` and a link to https://github.com/AMWA-TV/nmos-event-tally/blob/v1.0.x/docs/5.0.%20Transports.md#transports.
> 
> We can add that after this PR is approved and merged.

Yes that could be useful as a reminder for people to try both.

garethsb-sony commented on nmostesting/suites/IS0702Test.py:
```diff
@@ -95,24 +95,33 @@ def test_01(self, test):
+                                if self.sender_active_params[found_sender]["ext_is_07_source_id"] == source_id:
+                                    return test.PASS()
+                                else:
+                                    return test.FAIL("Source {} has no associated IS-05 sender".format(source_id))
```
How about `"IS-05 sender {} does not indicate the correct 'ext_is_07_source_id': {}".format(found_sender, source_id)`?

cristian-recoseanu replied:
Yep makes sense

garethsb-sony commented on nmostesting/suites/IS0702Test.py:
```diff
@@ -95,24 +95,33 @@ def test_01(self, test):
-                                if self.sender_active_params[sender]["ext_is_07_source_id"] == source:
-                                    found_source = True
+                                if self.sender_active_params[found_sender]["ext_is_07_source_id"] == source_id:
+                                    return test.PASS()
```
We can't `return` here, because we want to keep looping over the sources.

garethsb-sony replied:
And **oops**, I'd missed that this again only applies to WebSocket, we can't test MQTT senders this way.
For MQTT senders, we want to check `broker_topic` instead... but the format of that is only RECOMMENDED, so it can only be a WARNING if it doesn't match.

garethsb-sony commented on nmostesting/suites/IS0702Test.py:
```diff
@@ -95,24 +95,33 @@ def test_01(self, test):
                 else:
-                    return test.FAIL("Source {} not found in Node API".format(source))
-                return test.PASS()
+                    return test.FAIL("Source {} not found in Node API".format(source_id))
```
Based on above, I think we need to reinstate `return test.PASS()` here.

cristian-recoseanu commented:
Thanks for the improvement @garethsb-sony !
I have tried your latest commit against out implementations and it seems to work fine.
